src/Microsoft.AspNetCore.Authentication.Cookies/CookieAuthenticationHandler.cs

```diff
@@ -311,14 +311,27 @@ private CookieOptions BuildCookieOptions()
             Response.Headers[HeaderNames.CacheControl] = HeaderValueNoCache;
             Response.Headers[HeaderNames.Pragma] = HeaderValueNoCache;
             Response.Headers[HeaderNames.Expires] = HeaderValueMinusOne;
+
             if (shouldRedirectToReturnUrl && Response.StatusCode == 200)
             {
-                var query = Request.Query;
-                var redirectUri = query[Options.ReturnUrlParameter];
-                if (!StringValues.IsNullOrEmpty(redirectUri)
-                    && IsHostRelative(redirectUri))
+                CookieRedirectContext redirectContext = null;
+
+                // set redirect uri in order:
+                // 1. properties.RedirectUri
+                // 2. query parameter ReturnUrlParameter
+                var redirectUri = properties.RedirectUri;
+                if (string.IsNullOrEmpty(redirectUri) || !IsHostRelative(redirectUri))
```

Comment:
@Tratcher sure, query string values clearly fall into the NTUI category :smile:
My remark was only about the redirect URI stored in the authentication ticket.